crates/fluvio/src/producer/mod.rs

```diff
@@ -1,44 +1,331 @@
+            linger: Duration::from_millis(self.linger_ms.unwrap_or(DEFAULT_LINGER_MS)),
+            partitioner: self
+                .partitioner
+                .unwrap_or_else(|| Box::new(SiphashRoundRobinPartitioner::new())),
+        }
+    }
+}
+/// Options used to adjust the behavior of the Producer.
+/// Create this struct with `TopicProducerConfigBuilder`.
+///
+/// // Use config on this call:
+/// fluvio.topic_producer_with_config("topic", config).await?;
+/// # Ok(())
+/// # }
+/// ```
+pub struct TopicProducerConfig {
+    batch_size: usize,
+    linger: Duration,
+    partitioner: Box<dyn Partitioner + Send + Sync>,
+}
 
+impl Default for TopicProducerConfig {
+    fn default() -> Self {
+        Self {
+            linger: Duration::from_millis(DEFAULT_LINGER_MS),
+            batch_size: DEFAULT_BATCH_SIZE_BYTES,
+            partitioner: Box::new(SiphashRoundRobinPartitioner::new()),
+        }
+    }
+}
 /// An interface for producing events to a particular topic
 ///
 /// A `TopicProducer` allows you to send events to the specific
 /// topic it was initialized for. Once you have a `TopicProducer`,
 /// you can send events to the topic, choosing which partition
 /// each event should be delivered to.
 pub struct TopicProducer {
-    topic: String,
-    pool: Arc<SpuPool>,
-    partitioner: Box<dyn Partitioner + Send + Sync>,
+    inner: Arc<InnerTopicProducer>,
     #[cfg(feature = "smartengine")]
     pub(crate) smartmodule_instance: Option<Arc<RwLock<Box<dyn SmartModuleInstance>>>>,
 }
+
+use std::collections::VecDeque;
+struct PartitionProducer {
+    topic: String,
+    partition: PartitionId,
+    spu_pool: Arc<SpuPool>,
+    batches_lock: Arc<Mutex<VecDeque<ProducerBatch>>>,
+    linger: Duration,
+}
+
+impl PartitionProducer {
+    fn new(
+        topic: String,
+        partition: PartitionId,
+        spu_pool: Arc<SpuPool>,
+        batches_lock: Arc<Mutex<VecDeque<ProducerBatch>>>,
+        linger: Duration,
+    ) -> Self {
+        Self {
+            topic,
+            partition,
+            spu_pool,
+            batches_lock,
+            linger,
+        }
+    }
+
+    fn shared(
+        topic: String,
+        partition: PartitionId,
+        spu_pool: Arc<SpuPool>,
+        batches: Arc<Mutex<VecDeque<ProducerBatch>>>,
+        linger: Duration,
+    ) -> Arc<Self> {
+        Arc::new(PartitionProducer::new(
+            topic, partition, spu_pool, batches, linger,
+        ))
+    }
+
+    async fn flush(&self, force: bool) -> Result<()> {
+        let replica_key = ReplicaKey::new(self.topic.clone(), self.partition);
+        let partition_spec = self
+            .spu_pool
+            .metadata
+            .partitions()
+            .lookup_by_key(&replica_key)
+            .await?
+            .ok_or_else(|| FluvioError::PartitionNotFound(self.topic.to_string(), self.partition))?
+            .spec;
+        let leader = partition_spec.leader;
+
+        let spu_client = self
```

Comment:
why need to create a client at this point?

Reply:
We need to get the socket before sending the request, this under the hood does not create a new socket unless the previous one was staled